File: src/types/connection-profile.ts
// Re-export connection profile types from database.ts for better organization
export type {
  ConnectionProfile,
  AdvancedConnectionConfig,
  SSLConfig,
  ConnectionMetadata,
  ConnectionHealth,
  HealthCheckResult,
  ConnectionHealthHistory,
  MonitoringConfig,
  PoolStats,
  ConnectionMetrics,
  ExportData,
  EncryptedCredentials,
  ImportResult,
  ValidationResult,
  StorageStats,
  ProfileSearchFilters,
  ProfileSortOptions,
  ConnectionProfileManagerProps,
  QuickSelectorProps,
  HealthMonitorProps,
  ImportExportProps,
  AdvancedConfigProps,
  CreateConnectionProfileRequest,
  UpdateConnectionProfileRequest,
} from './database';

export {
  Environment,
  HealthStatus,
  MergeStrategy,
  DEFAULT_ADVANCED_CONFIG,
  DEFAULT_METADATA,
  DEFAULT_MONITORING_CONFIG,
} from './database';

// Import types for local use
import type {
  ConnectionProfile,
  ConnectionHealth,
  CreateConnectionProfileRequest,
} from './database';
import { Environment, HealthStatus } from './database';

// Additional utility types specific to connection profile management

export interface ConnectionProfileFormData {
  name: string;
  description: string;
  tags: string;
  folder: string;
  host: string;
  port: string;
  database: string;
  username: string;
  password: string;
  connectionTimeout: string;
  queryTimeout: string;
  maxConnections: string;
  idleTimeout: string;
  retryAttempts: string;
  retryDelay: string;
  sslMode: string;
  sslCert: string;
  sslKey: string;
  sslCa: string;
  customParameters: string;
  connectionStringTemplate: string;
  color: string;
  icon: string;
  isFavorite: boolean;
  autoConnect: boolean;
  environment: string;
  monitoringEnabled: boolean;
}

export interface ConnectionProfileValidation {
  isValid: boolean;
  errors: Record<string, string>;
  warnings: Record<string, string>;
}

export interface ConnectionTestResult {
  success: boolean;
  responseTimeMs?: number;
  errorMessage?: string;
  serverVersion?: string;
  databaseName?: string;
}

// ProfileSearchFilters and ProfileSortOptions are now imported from database.ts

export interface ConnectionProfileListItem {
  id: string;
  name: string;
  description?: string;
  tags: string[];
  folder?: string;
  host: string;
  database: string;
  environment: Environment;
  isFavorite: boolean;
  lastUsed?: Date;
  useCount: number;
  health?: ConnectionHealth;
}

// Form validation schemas (for use with form libraries like react-hook-form)
export interface ConnectionProfileFormSchema {
  name: {
    required: boolean;
    minLength: number;
    maxLength: number;
  };
  host: {
    required: boolean;
    pattern: RegExp;
  };
  port: {
    required: boolean;
    min: number;
    max: number;
  };
  database: {
    required: boolean;
    minLength: number;
  };
  username: {
    required: boolean;
    minLength: number;
  };
  password: {
    required: boolean;
    minLength: number;
  };
}

export const CONNECTION_PROFILE_FORM_SCHEMA: ConnectionProfileFormSchema = {
  name: {
    required: true,
    minLength: 1,
    maxLength: 100,
  },
  host: {
    required: true,
    pattern: /^[a-zA-Z0-9.-]+$/,
  },
  port: {
    required: true,
    min: 1,
    max: 65535,
  },
  database: {
    required: true,
    minLength: 1,
  },
  username: {
    required: true,
    minLength: 1,
  },
  password: {
    required: true,
    minLength: 1,
  },
};

// Constants for UI components
export const ENVIRONMENT_COLORS: Record<Environment, string> = {
  [Environment.Development]: 'bg-green-100 text-green-800 border-green-200',
  [Environment.Staging]: 'bg-yellow-100 text-yellow-800 border-yellow-200',
  [Environment.Production]: 'bg-red-100 text-red-800 border-red-200',
  [Environment.Testing]: 'bg-blue-100 text-blue-800 border-blue-200',
  [Environment.Other]: 'bg-gray-100 text-gray-800 border-gray-200',
};

export const HEALTH_STATUS_COLORS: Record<HealthStatus, string> = {
  [HealthStatus.Healthy]: 'text-green-600',
  [HealthStatus.Warning]: 'text-yellow-600',
  [HealthStatus.Error]: 'text-red-600',
  [HealthStatus.Unknown]: 'text-gray-600',
};

export const HEALTH_STATUS_ICONS: Record<HealthStatus, string> = {
  [HealthStatus.Healthy]: '●',
  [HealthStatus.Warning]: '⚠',
  [HealthStatus.Error]: '●',
  [HealthStatus.Unknown]: '○',
};

// Utility functions for type conversion
export function formDataToConnectionProfile(
  formData: ConnectionProfileFormData
): CreateConnectionProfileRequest {
  return {
    name: formData.name,
    description: formData.description || undefined,
    tags: formData.tags ? formData.tags.split(',').map(tag => tag.trim()) : [],
    folder: formData.folder || undefined,
    config: {
      host: formData.host,
      port: parseInt(formData.port, 10),
      database: formData.database,
      username: formData.username,
      connectionTimeout: parseInt(formData.connectionTimeout, 10),
      queryTimeout: parseInt(formData.queryTimeout, 10),
      maxConnections: parseInt(formData.maxConnections, 10),
      idleTimeout: parseInt(formData.idleTimeout, 10),
      retryAttempts: parseInt(formData.retryAttempts, 10),
      retryDelay: parseInt(formData.retryDelay, 10),
      sslConfig: {
        mode: formData.sslMode as any,
        cert: formData.sslCert || undefined,
        key: formData.sslKey || undefined,
        ca: formData.sslCa || undefined,
      },
      customParameters: formData.customParameters 
        ? JSON.parse(formData.customParameters) 
        : {},
      connectionStringTemplate: formData.connectionStringTemplate || undefined,
    },
    metadata: {
      color: formData.color || undefined,
      icon: formData.icon || undefined,
      isFavorite: formData.isFavorite,
      autoConnect: formData.autoConnect,
      environment: formData.environment as Environment,
      monitoringEnabled: formData.monitoringEnabled,
    },
  };
}

export function connectionProfileToFormData(
  profile: ConnectionProfile
): ConnectionProfileFormData {
  return {
    name: profile.name,
    description: profile.description || '',
    tags: profile.tags.join(', '),
    folder: profile.folder || '',
    host: profile.config.host,
    port: profile.config.port.toString(),
    database: profile.config.database,
    username: profile.config.username,
    password: '', // Never populate password in form
    connectionTimeout: profile.config.connectionTimeout.toString(),
    queryTimeout: profile.config.queryTimeout.toString(),
    maxConnections: profile.config.maxConnections.toString(),
    idleTimeout: profile.config.idleTimeout.toString(),
    retryAttempts: profile.config.retryAttempts.toString(),
    retryDelay: profile.config.retryDelay.toString(),
    sslMode: profile.config.sslConfig.mode,
    sslCert: profile.config.sslConfig.cert || '',
    sslKey: profile.config.sslConfig.key || '',
    sslCa: profile.config.sslConfig.ca || '',
    customParameters: JSON.stringify(profile.config.customParameters, null, 2),
    connectionStringTemplate: profile.config.connectionStringTemplate || '',
    color: profile.metadata.color || '',
    icon: profile.metadata.icon || '',
    isFavorite: profile.metadata.isFavorite,
    autoConnect: profile.metadata.autoConnect,
    environment: profile.metadata.environment,
    monitoringEnabled: profile.metadata.monitoringEnabled,
  };
}

export function validateConnectionProfile(
  formData: ConnectionProfileFormData
): ConnectionProfileValidation {
  const errors: Record<string, string> = {};
  const warnings: Record<string, string> = {};

  // Required field validation
  if (!formData.name.trim()) {
    errors.name = 'Connection name is required';
  }
  if (!formData.host.trim()) {
    errors.host = 'Host is required';
  }
  if (!formData.database.trim()) {
    errors.database = 'Database name is required';
  }
  if (!formData.username.trim()) {
    errors.username = 'Username is required';
  }
  if (!formData.password.trim()) {
    errors.password = 'Password is required';
  }

  // Port validation
  const port = parseInt(formData.port, 10);
  if (isNaN(port) || port < 1 || port > 65535) {
    errors.port = 'Port must be a number between 1 and 65535';
  }

  // Timeout validation
  const connectionTimeout = parseInt(formData.connectionTimeout, 10);
  if (isNaN(connectionTimeout) || connectionTimeout < 1) {
    errors.connectionTimeout = 'Connection timeout must be a positive number';
  }

  const queryTimeout = parseInt(formData.queryTimeout, 10);
  if (isNaN(queryTimeout) || queryTimeout < 1) {
    errors.queryTimeout = 'Query timeout must be a positive number';
  }

  // Connection pool validation
  const maxConnections = parseInt(formData.maxConnections, 10);
  if (isNaN(maxConnections) || maxConnections < 1) {
    errors.maxConnections = 'Max connections must be a positive number';
  }

  // Custom parameters validation
  if (formData.customParameters.trim()) {
    try {
      JSON.parse(formData.customParameters);
    } catch {
      errors.customParameters = 'Custom parameters must be valid JSON';
    }
  }

  // Warnings
  if (formData.environment === Environment.Production && !formData.sslMode.includes('require')) {
    warnings.sslMode = 'Consider using SSL for production connections';
  }

  if (maxConnections > 50) {
    warnings.maxConnections = 'High connection pool size may impact performance';
  }

  return {
    isValid: Object.keys(errors).length === 0,
    errors,
    warnings,
  };
}

File: src-tauri/src/connection_profile.rs
use chrono::{DateTime, Utc};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::time::Duration;

/// Core connection profile containing all information needed to manage a database connection
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionProfile {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub tags: Vec<String>,
    pub folder: Option<String>,
    pub config: AdvancedConnectionConfig,
    pub metadata: ConnectionMetadata,
    pub created_at: DateTime<Utc>,
    pub updated_at: DateTime<Utc>,
    pub last_used: Option<DateTime<Utc>>,
    pub use_count: u64,
}

/// Advanced connection configuration with all possible connection parameters
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct AdvancedConnectionConfig {
    // Basic connection info
    pub host: String,
    pub port: u16,
    pub database: String,
    pub username: String,
    
    // Advanced settings (serialized as seconds for Duration)
    #[serde(with = "duration_serde")]
    pub connection_timeout: Duration,
    #[serde(with = "duration_serde")]
    pub query_timeout: Duration,
    pub max_connections: u32,
    #[serde(with = "duration_serde")]
    pub idle_timeout: Duration,
    pub retry_attempts: u32,
    #[serde(with = "duration_serde")]
    pub retry_delay: Duration,
    
    // SSL Configuration
    pub ssl_config: SSLConfig,
    
    // Custom parameters
    pub custom_parameters: HashMap<String, String>,
    
    // Connection string template
    pub connection_string_template: Option<String>,
}

/// SSL/TLS configuration options
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SSLConfig {
    pub mode: SSLMode,
    pub cert: Option<String>,
    pub key: Option<String>,
    pub ca: Option<String>,
}

/// SSL modes supported by PostgreSQL
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum SSLMode {
    #[serde(rename = "disable")]
    Disable,
    #[serde(rename = "allow")]
    Allow,
    #[serde(rename = "prefer")]
    Prefer,
    #[serde(rename = "require")]
    Require,
    #[serde(rename = "verify-ca")]
    VerifyCa,
    #[serde(rename = "verify-full")]
    VerifyFull,
}

/// Metadata associated with a connection profile
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionMetadata {
    pub color: Option<String>,
    pub icon: Option<String>,
    pub is_favorite: bool,
    pub auto_connect: bool,
    pub environment: Environment,
    pub monitoring_enabled: bool,
}

/// Environment categorization for connections
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum Environment {
    #[serde(rename = "development")]
    Development,
    #[serde(rename = "staging")]
    Staging,
    #[serde(rename = "production")]
    Production,
    #[serde(rename = "testing")]
    Testing,
    #[serde(rename = "other")]
    Other(String),
}

/// Connection health status information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionHealth {
    pub status: HealthStatus,
    pub last_checked: DateTime<Utc>,
    pub response_time_ms: Option<u64>,
    pub error_message: Option<String>,
}

/// Health status enumeration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum HealthStatus {
    #[serde(rename = "healthy")]
    Healthy,
    #[serde(rename = "warning")]
    Warning,
    #[serde(rename = "error")]
    Error,
    #[serde(rename = "unknown")]
    Unknown,
}

/// Health check result with timestamp
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct HealthCheckResult {
    pub timestamp: DateTime<Utc>,
    pub status: HealthStatus,
    pub response_time_ms: Option<u64>,
    pub error_message: Option<String>,
}

/// Connection health history
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionHealthHistory {
    pub current: ConnectionHealth,
    pub history: Vec<HealthCheckResult>,
    pub uptime_percentage: f64,
}

/// Monitoring configuration
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct MonitoringConfig {
    pub enable_auto_check: bool,
    pub check_interval_minutes: u32,
    pub enable_notifications: bool,
    pub critical_connection_ids: Vec<String>,
}

/// Pool statistics for connection monitoring
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PoolStats {
    pub active_connections: u32,
    pub idle_connections: u32,
    pub max_connections: u32,
    pub total_connections_created: u64,
    pub average_wait_time_ms: f64,
}

/// Connection metrics for monitoring
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ConnectionMetrics {
    pub total_profiles: u32,
    pub active_connections: u32,
    pub average_response_time_ms: f64,
    pub success_rate: f64,
    pub uptime_percentage: f64,
}

/// Merge strategy for importing profiles
#[derive(Debug, Clone, Serialize, Deserialize)]
pub enum MergeStrategy {
    #[serde(rename = "replace")]
    Replace,
    #[serde(rename = "merge")]
    Merge,
    #[serde(rename = "skip")]
    Skip,
}

/// Export data structure
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ExportData {
    pub version: String,
    pub exported_at: DateTime<Utc>,
    pub profiles: Vec<ConnectionProfile>,
    pub credentials: Option<HashMap<String, EncryptedCredentials>>,
    pub checksum: String,
}

/// Encrypted credentials for secure storage
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct EncryptedCredentials {
    pub encrypted_password: Vec<u8>,
    pub nonce: Vec<u8>,
    pub encrypted_at: DateTime<Utc>,
}

/// Import result information
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ImportResult {
    pub imported_count: u32,
    pub skipped_count: u32,
    pub error_count: u32,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
}

/// Validation result for import data
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct ValidationResult {
    pub is_valid: bool,
    pub version_compatible: bool,
    pub profile_count: u32,
    pub errors: Vec<String>,
    pub warnings: Vec<String>,
}

// Custom serialization for Duration to handle JSON serialization
mod duration_serde {
    use serde::{Deserialize, Deserializer, Serialize, Serializer};
    use std::time::Duration;

    pub fn serialize<S>(duration: &Duration, serializer: S) -> Result<S::Ok, S::Error>
    where
        S: Serializer,
    {
        duration.as_secs().serialize(serializer)
    }

    pub fn deserialize<'de, D>(deserializer: D) -> Result<Duration, D::Error>
    where
        D: Deserializer<'de>,
    {
        let secs = u64::deserialize(deserializer)?;
        Ok(Duration::from_secs(secs))
    }
}

impl Default for AdvancedConnectionConfig {
    fn default() -> Self {
        Self {
            host: "localhost".to_string(),
            port: 5432,
            database: "postgres".to_string(),
            username: "postgres".to_string(),
            connection_timeout: Duration::from_secs(30),
            query_timeout: Duration::from_secs(300), // 5 minutes
            max_connections: 10,
            idle_timeout: Duration::from_secs(300), // 5 minutes
            retry_attempts: 3,
            retry_delay: Duration::from_secs(1),
            ssl_config: SSLConfig::default(),
            custom_parameters: HashMap::new(),
            connection_string_template: None,
        }
    }
}

impl Default for SSLConfig {
    fn default() -> Self {
        Self {
            mode: SSLMode::Prefer,
            cert: None,
            key: None,
            ca: None,
        }
    }
}

impl Default for ConnectionMetadata {
    fn default() -> Self {
        Self {
            color: None,
            icon: None,
            is_favorite: false,
            auto_connect: false,
            environment: Environment::Development,
            monitoring_enabled: false,
        }
    }
}

impl Default for MonitoringConfig {
    fn default() -> Self {
        Self {
            enable_auto_check: false,
            check_interval_minutes: 5,
            enable_notifications: true,
            critical_connection_ids: Vec::new(),
        }
    }
}

impl ConnectionProfile {
    /// Create a new connection profile with default values
    pub fn new(name: String, config: AdvancedConnectionConfig) -> Self {
        let now = Utc::now();
        Self {
            id: uuid::Uuid::new_v4().to_string(),
            name,
            description: None,
            tags: Vec::new(),
            folder: None,
            config,
            metadata: ConnectionMetadata::default(),
            created_at: now,
            updated_at: now,
            last_used: None,
            use_count: 0,
        }
    }

    /// Update the last used timestamp and increment use count
    pub fn mark_used(&mut self) {
        self.last_used = Some(Utc::now());
        self.use_count += 1;
        self.updated_at = Utc::now();
    }

    /// Update the profile and set updated_at timestamp
    pub fn update(&mut self) {
        self.updated_at = Utc::now();
    }
}

impl AdvancedConnectionConfig {
    /// Convert to a PostgreSQL connection string
    pub fn to_connection_string(&self, password: &str) -> String {
        if let Some(template) = &self.connection_string_template {
            // Use custom template if provided
            template
                .replace("{host}", &self.host)
                .replace("{port}", &self.port.to_string())
                .replace("{database}", &self.database)
                .replace("{username}", &self.username)
                .replace("{password}", password)
        } else {
            // Build standard PostgreSQL connection string
            let mut conn_str = format!(
                "postgresql://{}:{}@{}:{}/{}",
                self.username, password, self.host, self.port, self.database
            );

            // Add SSL mode
            conn_str.push_str(&format!("?sslmode={}", self.ssl_config.mode.to_string()));

            // Add connection timeout
            conn_str.push_str(&format!("&connect_timeout={}", self.connection_timeout.as_secs()));

            // Add custom parameters
            for (key, value) in &self.custom_parameters {
                conn_str.push_str(&format!("&{}={}", key, value));
            }

            conn_str
        }
    }
}

impl SSLMode {
    pub fn to_string(&self) -> &'static str {
        match self {
            SSLMode::Disable => "disable",
            SSLMode::Allow => "allow",
            SSLMode::Prefer => "prefer",
            SSLMode::Require => "require",
            SSLMode::VerifyCa => "verify-ca",
            SSLMode::VerifyFull => "verify-full",
        }
    }
}

impl Environment {
    pub fn to_string(&self) -> String {
        match self {
            Environment::Development => "development".to_string(),
            Environment::Staging => "staging".to_string(),
            Environment::Production => "production".to_string(),
            Environment::Testing => "testing".to_string(),
            Environment::Other(name) => name.clone(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_connection_profile_creation() {
        let config = AdvancedConnectionConfig::default();
        let profile = ConnectionProfile::new("Test Connection".to_string(), config);
        
        assert_eq!(profile.name, "Test Connection");
        assert_eq!(profile.use_count, 0);
        assert!(profile.last_used.is_none());
        assert!(!profile.id.is_empty());
    }

    #[test]
    fn test_connection_profile_mark_used() {
        let config = AdvancedConnectionConfig::default();
        let mut profile = ConnectionProfile::new("Test Connection".to_string(), config);
        
        profile.mark_used();
        
        assert_eq!(profile.use_count, 1);
        assert!(profile.last_used.is_some());
    }

    #[test]
    fn test_connection_string_generation() {
        let config = AdvancedConnectionConfig {
            host: "localhost".to_string(),
            port: 5432,
            database: "testdb".to_string(),
            username: "testuser".to_string(),
            ..Default::default()
        };
        
        let conn_str = config.to_connection_string("testpass");
        
        assert!(conn_str.contains("postgresql://testuser:testpass@localhost:5432/testdb"));
        assert!(conn_str.contains("sslmode=prefer"));
        assert!(conn_str.contains("connect_timeout=30"));
    }

    #[test]
    fn test_ssl_mode_serialization() {
        let ssl_config = SSLConfig {
            mode: SSLMode::Require,
            cert: Some("cert.pem".to_string()),
            key: Some("key.pem".to_string()),
            ca: Some("ca.pem".to_string()),
        };
        
        let json = serde_json::to_string(&ssl_config).unwrap();
        let deserialized: SSLConfig = serde_json::from_str(&json).unwrap();
        
        assert_eq!(ssl_config.mode.to_string(), deserialized.mode.to_string());
        assert_eq!(ssl_config.cert, deserialized.cert);
    }

    #[test]
    fn test_duration_serialization() {
        let config = AdvancedConnectionConfig::default();
        
        let json = serde_json::to_string(&config).unwrap();
        let deserialized: AdvancedConnectionConfig = serde_json::from_str(&json).unwrap();
        
        assert_eq!(config.connection_timeout, deserialized.connection_timeout);
        assert_eq!(config.query_timeout, deserialized.query_timeout);
    }
}